local-volume/provisioner/pkg/monitor/monitor.go

```diff
@@ -0,0 +1,479 @@
+	}
+
+	monitor.localVolumeMap = NewLocalVolumeMap()
+
+	monitor.volumeLW = &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return config.Client.CoreV1().PersistentVolumes().List(labelOps)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return config.Client.CoreV1().PersistentVolumes().Watch(labelOps)
+		},
+	}
+	_, monitor.volumeController = cache.NewInformer(
+		monitor.volumeLW,
+		&v1.PersistentVolume{},
+		DefaultInformerResyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    monitor.addVolume,
+			UpdateFunc: monitor.updateVolume,
+			DeleteFunc: monitor.deleteVolume,
+		},
+	)
+
+	// fill map at first with data from ETCD
+	monitor.flushFromETCDFirst()
+
+	return monitor
+}
+
+// flushFromETCDFirst fill map with data from etcd at first
+func (monitor *Monitor) flushFromETCDFirst() error {
+	pvs, err := monitor.Client.CoreV1().PersistentVolumes().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	if len(pvs.Items) == 0 {
+		glog.Infof("no pv in ETCD at first")
+		return nil
+	}
+
+	for _, pv := range pvs.Items {
+		monitor.localVolumeMap.AddLocalVolume(&pv)
+	}
+	return nil
+}
+
+func (monitor *Monitor) addVolume(obj interface{}) {
+	volume, ok := obj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", obj)
+		return
+	}
+
+	monitor.localVolumeMap.AddLocalVolume(volume)
+
+}
+
+func (monitor *Monitor) updateVolume(oldObj, newObj interface{}) {
+	newVolume, ok := newObj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", newObj)
+		return
+	}
+
+	monitor.localVolumeMap.UpdateLocalVolume(newVolume)
+}
+
+func (monitor *Monitor) deleteVolume(obj interface{}) {
+	volume, ok := obj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", obj)
+		return
+	}
+
+	monitor.localVolumeMap.DeleteLocalVolume(volume)
+
+}
+
+// Run starts all of this controller's control loops
+func (monitor *Monitor) Run(stopCh <-chan struct{}) {
+	glog.Infof("Starting monitor controller %s!", string(monitor.RuntimeConfig.Name))
+	monitor.hasRunLock.Lock()
+	monitor.hasRun = true
+	monitor.hasRunLock.Unlock()
+	go monitor.volumeController.Run(stopCh)
+
+	go monitor.MonitorLocalVolumes()
+	<-stopCh
+}
+
+// HasRun returns whether the volume controller has Run
+func (monitor *Monitor) HasRun() bool {
+	monitor.hasRunLock.Lock()
+	defer monitor.hasRunLock.Unlock()
+	return monitor.hasRun
+}
+
+// MonitorLocalVolumes checks local PVs periodically
+func (monitor *Monitor) MonitorLocalVolumes() {
+	for {
+		if monitor.HasRun() {
+			pvs := monitor.localVolumeMap.GetPVs()
+			for _, pv := range pvs {
+				monitor.checkStatus(pv)
+			}
+		}
+
+		time.Sleep(DefaultMonitorResyncPeriod)
+	}
+}
+
+// checkStatus checks pv health condition
+func (monitor *Monitor) checkStatus(pv *v1.PersistentVolume) {
```

Comment:
Do we need to use `fsck` or `smartctl` to check the disk healthy?

Reply:
yes, this is just a simple demo, lots of extra work needs to be done for local pv monitor.

Reply:
and i am updating the proposal now, will send it out when ready

Reply:
@NickrenREN Thanks. Looking forward to the Proposal.